google/cloud/forseti/scanner/audit/role_rules_engine.py

```diff
@@ -0,0 +1,266 @@
+                    res=res)
+
+        if role_name not in self.rules_map:
+            self.rules_map[role_name] = [rule]
+        else:
+            self.rules_map[role_name].append(rule)
+
+    def get_rule_by_role_name(self, role_name):
+        """Get the rule of a given role.
+
+        Args:
+            role_name (str): Name of a role.
+        Returns:
+            list: Rule list of the given role.
+        """
+        return self.rules_map.get(role_name, [])
+
+
+class Rule(object):
+    """Rule properties from the rule definition file. Also finds violations."""
+
+    def __init__(self, rule_index, rule_name, permissions, res):
+        """Initialize.
+
+        Args:
+            rule_index (int): The index of the rule.
+            rule_name (str): Name of the rule.
+            permissions (int): Expected permissions of the role.
+            res (dict): Parent resource of the role that should obey the rule.
+        """
+        self.rule_name = rule_name
+        self.rule_index = rule_index
+        self.permissions = permissions[:]
+        self.res_types = res[:]
+
+        for res_item in self.res_types:
+            if 'type' not in res_item:
+                raise audit_errors.InvalidRulesSchemaError(
+                    'Lack of resource:type in rule {}'.format(rule_index))
+            if 'resource_ids' not in res_item:
+                raise audit_errors.InvalidRulesSchemaError(
+                    'Lack of resource:resource_ids in rule {}'.format(
+                        rule_index))
+
+            if '*' in res_item['resource_ids']:
+                res_item = ['*']
+
+    def generate_violation(self, role):
+        """Generate a violation.
+
+        Args:
+            role (Role): The role that triggers the violation.
+        Returns:
+            RuleViolation: The violation.
+        """
+        permissions = role.get_permissions()
+        permissions_str = json.dumps(permissions)
+
+        return RuleViolation(
+            resource_name=role.name,
+            resource_id=role.id,
+            resource_type=role.type,
+            full_name=role.full_name,
+            rule_name=self.rule_name,
+            rule_index=self.rule_index,
+            violation_type=VIOLATION_TYPE,
+            violation_data=permissions_str,
+            resource_data=role.data,
+        )
+
+    def find_violations(self, res):
+        """Get a generator for violations.
+
+        Args:
+            res (Resource): A class derived from Resource.
+        Returns:
+            Generator: All violations of the resource breaking the rule.
+
+        Raises:
+            ValueError: Raised if the resource type is bucket.
+        """
+
+        if res.type == 'role':
+            return self.find_violations_in_role(res)
+        raise ValueError(
+            'only role is supported.'
+        )
+
+    def find_violations_in_role(self, role):
+        """Get a generator for violations.
```

Comment:
This scanner will only look at 'Rule' resources, so you can move the body of this function into `find_violations()`